source/logrepl/internal/subscription.go

```diff
@@ -315,20 +320,33 @@ func (s *Subscription) sendStandbyStatusUpdate(ctx context.Context) error {
 		return fmt.Errorf("walWrite (%s) should be >= walFlush (%s)", s.walWritten, walFlushed)
 	}
 
+	// N.B. Manage replication slot lag, by responding with the last server LSN, when
+	//      all previous slot relevant msgs have been written and flushed
+	replyWithWALEnd := walFlushed == s.walWritten && walFlushed < s.serverWALEnd
+
 	sdk.Logger(ctx).Trace().
-		Str("walWrite", s.walWritten.String()).
-		Str("walFlush", walFlushed.String()).
-		Str("walApply", walFlushed.String()).
+		Stringer("wal_write", s.walWritten).
+		Stringer("wal_flush", walFlushed).
+		Stringer("server_wal_end", s.serverWALEnd).
+		Bool("server_wal_end_sent", replyWithWALEnd).
 		Msg("sending standby status update")
 
-	err := pglogrepl.SendStandbyStatusUpdate(ctx, s.conn, pglogrepl.StandbyStatusUpdate{
+	if replyWithWALEnd {
+		if err := pglogrepl.SendStandbyStatusUpdate(ctx, s.conn, pglogrepl.StandbyStatusUpdate{
```

Comment:
There's an oopsie - if `replyWithWALEnd` is `true` we will send two status updates.

Rather do something like this:
```suggestion
	var statusUpdate pglogrepl.StandbyStatusUpdate
	if replyWithWALEnd {
		statusUpdate = pglogrepl.StandbyStatusUpdate{...}
	} else {
		statusUpdate = pglogrepl.StandbyStatusUpdate{...}
	}
	pglogrepl.SendStandbyStatusUpdate(ctx, s.conn, statusUpdate)
```

Reply:
Oh yes fixing.

Reply:
Fixed in 900b921e8b6ee400e970389aa11b107b9d26d8ca